rfc-175-frontend-fewer-apps.md

```diff
@@ -0,0 +1,254 @@
+
+We propose using [frontend] as the host app. This is because:
+
+* `frontend` contains some of the more complicated document types like licence_transaction and local_transaction, that have multiple pages and interactive elements, which would be harder to consolidate into another app
+* `frontend` has a one-controller per document type layout, which is preferable for document types with multiple pages as it follows [rails conventions for controllers] to do the orchestration.
+
+[government-frontend] was also considered for the host app, but not chosen because it has a more restrictive presenter pattern that works well for single-page document types. Even though `government-frontend` has many document types and accounts for 50% of the frontend traffic, it would be easier to move the single-page document types from `government-frontend` to `frontend` than it would to try and force the document types in `frontend` into the presenter pattern.
+
+### Testing standardisation
+We propose standardising around RSpec as the test suite for the frontend, ideally [modernising the tests], for example by switching to request tests rather than controller tests and system tests rather than feature tests. RSpec is the [preferred testing suite] on GOV.UK and is favoured by many developers because the tests and testing output are easier to read. Of the 4 applications only [government-frontend] uses minitest at this point, so its test suite could be converted route-by-route as it is merged into [frontend]
+
+### Code structure
+The code structure will be standardised. As well as using the [default Rails directory structure], we propose the following standard for additional classes (e.g. services, api calls). The standard expands on the [conventions] documented in the developer docs.
+
+The host app will follow the default MVC structure of Rails applications, with some additional restrictions on what goes where that are frontend specific (listed below)
+
+```
+├── app
+│   ├── assets
+│   ├── controllers
+│   ├── helpers
+│   ├── models
+│   ├── presenters
+│   ├── views
+├── config
+│   ├── data
+│   ├── locales
+├── lib
+│   ├── tasks
+└── .gitignore
+```
+
+Under `/app`:
+
+`helpers`
+- General utility methods for a view.
+
+`models`
+- Frontend apps do not have database records, so models here are created to store data received from api calls (eg content items from content store), process data that will be written to apis (eg feedback items to zendesk), and information ingested from YAML/JSON/CSVs.
+
+`presenters`
+- Presentation related transformations on model data. It is a wrapper around a model for use in a view.
+
+Under `/lib`
+
+`tasks`
+- Rake tasks
+
+Use `lib` rather than `app/lib` for custom code that enhances the app but doesn’t fit into a directory in `app`
+
+Under `/config`
+
+`data`
+- YAML, JSON, CSV config files. Data files should be placed here rather than under `/lib/data`.
```

Comment:
In the time since the last RFC we on the AI team have had to deal in anger with considering the place for `data` files and found the `config/data` route was actually a bit surprising and `lib/data` was a better choice.

Our rationale was:

1. when people tend to be reaching for a `data` directory it tends to not actually be things that configure the app - so config/data seemed surprising for something that isn't doing any configuration
2. when putting things into the config directory you would typically want them to be accessible via `Rails.configuration` to be conventional config, and it seems strange to potentially have a directory of files that isn't
3. Rails already has the precedent for non namespaced storage in lib, with the assets and tasks precedence. It's really easy to flag a directory to be ignored by the autoloader: https://github.com/alphagov/govuk-chat/blob/54059d50aff6211328fb4f8302333adaaa03b02d/config/application.rb#L26-L29

The tricky bit is the blurring of the line between what is config and what is data - which you'll probably have even with this decision to choose between whether a file lives in `config` or `config/data`. The barometer we found worked well was whether it makes sense to be part of `Rails.configuration`

Reply:
Hmm. Looking at the current use of lib/data, only frontend uses it really (two apps use it for recruitment banner config, but as per RFC-176 we're looking to remove that code in app anyway), and it uses it to hold two content items that it publishes. (bank holiday and clock change content items). Given that those should probably not be in Frontend anyway, I think this might be a moot point in practice. In theory, though, we could have things in both directories depending on what they were?

Reply:
> whether it makes sense to be part of `Rails.configuration`

I think that statement is open to a lot of interpretation too.
There doesn't seem to be a consensus in the wider world about where to data files either. It might be better to have a top level `/data` directory for model data.

Reply:
I've added a note here that /config/data should be for data that's clearly for configuration, a root data directory `/data` for model data, but that in general we should try to avoid using data files for configuration in these apps.